decomp/semantics/uds.py

```diff
@@ -960,37 +1188,199 @@ def _add_edge_annotation(self, edge, attrs):
             warning(warnmsg)
             self.graph.add_edges_from([(edge[0], edge[1], attrs)])
 
+    @memoized_property
+    def sentence(self) -> str:
+        """The sentence annotated by this graph"""
+        id_word = {nodeattr['position']-1: nodeattr['form']
+                   for nodeid, nodeattr in self.syntax_nodes.items()}
+        return ' '.join([id_word[i] for i in range(max(list(id_word.keys()))+1)])
+
+
+class UDSDocumentGraph(UDSGraph):
+    """A Universal Decompositional Semantics document-level graph
+
+    Parameters
+    ----------
+    graph
+        the NetworkX DiGraph from which the document-level graph
+        is to be constructed
+    name
+        the name of the graph
+    """
+    @overrides
+    def __init__(self, graph: DiGraph, name: str):
+        super().__init__(graph, name)
+
+    def add_annotation(self,
+                   node_attrs: Dict[str, Dict[str, Any]],
+                   edge_attrs: Dict[str, Dict[str, Any]]) -> None:
+        """Add node and or edge annotations to the graph
+
+        Parameters
+        ----------
+        node_attrs
+            the node annotations to be added
+        edge_attrs
+            the edge annotations to be added
+        """
+        for node, attrs in node_attrs.items():
+            self._add_node_annotation(node, attrs)
+
+        for edge, attrs in edge_attrs.items():
+            self._add_edge_annotation(edge, attrs)
+
+    def _add_edge_annotation(self, edge, attrs):
+        if edge in self.graph.edges:
+            self.graph.edges[edge].update(attrs)
+        else:
+            self.graph.add_edges_from([(edge[0], edge[1], attrs)])
+
+    def _add_node_annotation(self, node, attrs):
+        # We do not currently have a use case for document node annotations,
+        # but it is included for completeness.
+        if node in self.graph.nodes:
+            warnmsg = f'Attempting to add a node annotation to node {node} '\
```

Comment:
Avoid f-strings for users of Python < 3.6.

Reply:
I think we discussed this in my [last PR](https://github.com/decompositional-semantics-initiative/decomp/pull/7#discussion_r446952694). The Dockerfile requires >= 3.6, but if someone's not using Docker, you're right. I can replace all the f-strings.

Reply:
I remember that discussion, but I misremembered that we decided against using them. Leaving them in is fine.